radicle-tracker/src/ops/id.rs

```diff
@@ -0,0 +1,113 @@
+// This file is part of radicle-link
+// <https://github.com/radicle-dev/radicle-link>
+//
+// Copyright (C) 2019-2020 The Radicle Team <dev@radicle.xyz>
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 or
+// later as published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use crate::metadata::clock::TimeDiff;
+use std::{cmp::Ordering, fmt, hash::Hash, time::SystemTime};
+use uuid::Uuid;
+
+/// Magically generate a thing.
+pub trait Gen {
+    /// Abrakedabra!
+    fn gen() -> Self;
+}
+
+/// A unique identifier consisting of a blob of bytes (for some meaning of
+/// unique).
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Unique {
+    blob: Vec<u8>,
+}
+
+impl Unique {
+    /// Peek at the bytes of the unique identifier.
+    pub fn val(&self) -> &[u8] {
+        &self.blob
+    }
+}
+
+impl Gen for Unique {
+    fn gen() -> Self {
+        Unique {
+            blob: Uuid::new_v4().as_bytes().to_vec(),
+        }
+    }
+}
+
+impl fmt::Display for Unique {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let utf8 = String::from_utf8_lossy(&self.blob);
+        write!(f, "{}", utf8)
+    }
+}
+
+/// A combination of [`Unique`] identifier along with a [`TimeDiff`].
+///
+/// The ordering of a `UniqueTimestamp` first relies on its `TimeDiff` falling
+/// back to the identifier if the times were equal.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct UniqueTimestamp {
+    unique: Unique,
+    time: TimeDiff,
+}
+
+impl PartialOrd for UniqueTimestamp {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(&other))
+    }
+}
+
+impl Ord for UniqueTimestamp {
+    fn cmp(&self, other: &Self) -> Ordering {
+        match self.time.cmp(&other.time) {
```

Comment:
Why not also comparing `unique` if times are equal?
This would make ordering total (only equal `UniqueTimestamp`s would be... equal), and would allow using BTreeMap as backing store.

Reply:
> Why not also comparing unique if times are equal?

But... it does? :smile: 

>  This would make ordering total (only equal UniqueTimestamps would be... equal), and would allow using BTreeMap as backing store.

I tried this on a separate branch. I wasn't happy with the ergonomics from a user point of view. It gets hard to say: I want to react to this reply, because you have to get the `UniqueTimestamp` to reference it. Comparing that to just using `usize` as indices. But I'm happy to create a ticket to explore it again :) Does that sound good to you?